pallets/transfer-allowlist/src/lib.rs

```diff
@@ -0,0 +1,533 @@
+			sender_account_id: T::AccountId,
+			currency_id: T::CurrencyId,
+			delay: T::BlockNumber,
+		},
+		/// Event for Allowance delay removal
+		TransferAllowanceDelayRemoval {
+			sender_account_id: T::AccountId,
+			currency_id: T::CurrencyId,
+		},
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Adds a transfer allowance for a sending Account/Currency.
+		/// Allowance either starts at the current block + the delay set for the account,
+		/// if a delay is present.
+		/// or block 0 if no delay is present.
+		/// Important! Account/Currency sets with an allowance set are restricted to just the allowances added for the account -
+		/// to have unrestricted transfers allowed for the sending Account and Currency, no allowances should be present.
+		///
+		/// Running this for an existing allowance generates a new allowance based on the current delay, or lack thereof
+		#[pallet::call_index(0)]
+		#[pallet::weight(10_000 + T::DbWeight::get().reads_writes(2, 2).ref_time())]
+		pub fn add_transfer_allowance(
+			origin: OriginFor<T>,
+			currency_id: T::CurrencyId,
+			receiver: T::Location,
+		) -> DispatchResult {
+			let account_id = ensure_signed(origin)?;
+
+			let allowance_details = match Self::get_account_currency_restriction_count_delay(
+				&account_id,
+				currency_id,
+			) {
+				Some((_, Some(delay))) => AllowanceDetails {
+					allowed_at: <frame_system::Pallet<T>>::block_number().saturating_add(delay),
+					..AllowanceDetails::default()
+				},
+				_ => AllowanceDetails::default(),
+			};
+			if !<AccountCurrencyTransferAllowance<T>>::contains_key((
+				&account_id,
+				&currency_id,
+				&receiver,
+			)) {
+				Self::increment_or_create_allowance_count(&account_id, &currency_id)?;
+				T::ReserveCurrency::reserve(
+					&account_id,
+					T::Fees::fee_value(T::AllowanceFeeKey::get()),
+				)?;
+			};
+			<AccountCurrencyTransferAllowance<T>>::insert(
+				(&account_id, &currency_id, &receiver),
+				&allowance_details,
+			);
+
+			Self::deposit_event(Event::TransferAllowanceCreated {
+				sender_account_id: account_id,
+				currency_id,
+				receiver,
+				allowed_at: allowance_details.allowed_at,
+				blocked_at: allowance_details.blocked_at,
+			});
+			Ok(())
+		}
+
+		/// Restricts a transfer allowance for a sending account/currency/receiver location to:
+		/// - either the current block + delay if a delay is set
+		/// - or the current block if no delay is set
+		#[pallet::call_index(1)]
+		#[pallet::weight(10_000 + T::DbWeight::get().reads_writes(2, 2).ref_time())]
+		pub fn remove_transfer_allowance(
+			origin: OriginFor<T>,
+			currency_id: T::CurrencyId,
+			receiver: T::Location,
+		) -> DispatchResult {
+			let account_id = ensure_signed(origin)?;
+
+			let blocked_at = match Self::get_account_currency_restriction_count_delay(
+				&account_id,
+				currency_id,
+			) {
+				Some((_, Some(delay))) => {
+					<frame_system::Pallet<T>>::block_number().saturating_add(delay)
```

Comment:
I was thinking if it might makes to have two values here. 
* One for delay of adding allowance
* One for delay of removing allowance

cc @offerijns do you think this will be an issue for issuers?

Reply:
This would introduce quite some rewrite I guess, so I only want it if we think it makes sense. Thinking from a security perspective here. But actually, if a key is compromised and the holder did not set a delay upfront, then the attacker can do anything directly anyways. If a delay was set, then fast removal would not introduce any advantage. WDYT?

Reply:
I'm not sure it would make sense strictly from a security perspective, however I could see it from a permissions and  convienience with fee reclaiming perspective.  I'm open to however we want to handle this :)

Reply:
Also, would we want to add another reserve for storage usage for the delay?

Reply:
Mmh. Good point. I think it makes sense. But I am not sure, if it is not a little bit hard for users to follow how they can free it it up if we have lots of little locations. 

But I am nto against having an additional fee.

Reply:
Missed seeing this earlier unfortunately.  I would prefer keeping no fee for simplicity, but I'm wondering now if having users be able to add a delay without a fee could be a DOS risk, or would the transaction/weight fee be enough of a deterrent for that?